README.md
# GMCalc-EVO
This repository includes the Modified GMCalc according to my own usage based on v1.5.3. The original GMCalc can be found in https://people.physics.carleton.ca/~logan/gmcalc/


CHANGES.txt
GMCALC -- changes log

2022/08/04 - version 1.5.3:
	     Fixed bugs in src/gmhbhs.f that affected HiggsBounds results
	     when h, H, or H30 masses are close to the ttbar threshold.

2022/07/04 - version 1.5.2:
	     Modified vegas.f to comply with fortran 2018 standards
	     (NO change to the algorithm).
	     Tested with LoopTools 2.16.

2022/06/04 - version 1.5.1:
	     Updated CMS direct search for VBF -> H5++/-- (2104.04762).
	     Incorporated new ATLAS direct search for Drell-Yan H++H-- 
	     -> 4W (2101.11961).

2021/02/05 - version 1.5.0:
             Added interface to call HiggsBounds-5 and HiggsSignals-2.
	     Added routines to apply direct constraints from VBF -> H5++, 
             Drell-Yan H5++, and Drell-Yan H50 --> gamma gamma, which 
	     are not in HiggsBounds 5.3.0.
             Rewrote calculation of MSQ22 in READIN2 to properly handle 
             v_chi = 0 case.

2018/09/30 - version 1.4.1 (new inputset):
	     Added INPUTSET 6 (mh, m5, sin(thetaH), lambda2, lambda3,
	     lambda4, lambda5, M2).

2018/07/25 - version 1.4.0: 
	     gmmg5.f now also generates a param_card for the (LO) 
	     UFO model file with effective vertices for loop-induced 
	     Higgs decays.
	     Improved integration speed for scalar decay widths to 
	     doubly-offshell vector boson pairs.
             Updated makefile to work with LoopTools 2.15.
	     Updated S and T parameters to PDG2018 values.
	     Bug fixed in H3+ --> W- H5++ width calculation.
	     Offshell H1 -> V* H2 set to zero when V -> H1 H2 is 
	     possible: formula used was not valid in that region.

2017/08/30 - version 1.3.0:
	     One-loop decays of H5+ --> W+ gamma, H3+ --> W+ gamma,
	     and H50 --> Z gamma are now computed as in C.Degrande, 
	     K.Hartling & H.E.Logan, arXiv:1708.08753.  Computation of 
	     these decays requires linking LoopTools; see 
	     www.feynarts.de/looptools/ .  See the makefile for compilation
	     without LoopTools, i.e., setting these decays to zero.

2016/06/20 - version 1.2.1 (new inputset):
             Added INPUTSET 5 (mh, mH, sin(thetaH), sin(alpha), lambda2,
             lambda3, lambda4, lambda5).

2015/11/30 - version 1.2.0:
             Doubly-offshell H -> V*V* decays implemented for all masses;
	     numerics validated using HDECAY 6.42 to be good to within 1%.
             New inputs GAMZ and GAMW are the physical widths of the Z and W.
	     gmmg5.f updated to produce LO and NLO param-cards for use with 
	     the corresponding UFO model files.  Widths set to Auto so that
	     they will be calculated by MadGraph.

2015/08/11 - version 1.1.1 (bug fixes):
	     Bugs fixed in h, H, H3^0, and H5^0 decays to Z gamma - thanks
	     to Celine Degrande.

2015/03/26 - version 1.1.0: 
	     Added INPUTSET 4 (mh, m5, sH, lambda2, lambda3, lambda4, M1, M2).
	     Changed default EW input parameters to GF, MZ, MW.
	     Added consistency check for INPUTSET 3.

2015/01/20 - version 1.0.1 (bug fixes):
	     Fixed the handling of a singular case in the routine that finds
	     roots of the cubic equation for VCHI in gmspectrum.f - thanks 
  	     to R. Ruiz for pointing out this bug.
	     Fixed the code that chooses VCHI positive in gmscan.f.

2014/12/19 - added INPUTSET 3 (physical masses and mixing angles)
2014/12/15 - changed variable name in param_card.dat (to match FeynRules)
2014/11/12 - bug fixed in GET_MINMASS in src/gmutils.f
2014/11/07 - first public version of GMCALC v1.0


src/vegas.f
C==================================================================
C  EVERYTHING BELOW THIS LINE BELONGS TO VEGAS
C  AND WAS COPIED FROM PROSPINO 2002/02/22.
C==================================================================


************************************************************************
*                                                                      *
*     SUBROUTINE INTEG (CALLING VEGAS):                                *
*     ---------------------------------                                *
*                                                                      *
*     IDIM-DIMENSIONAL MONTE CARLO INTEGRATION OF FUNCTION "FUNCT"     *
*     1ST RUN (VEGAS) : IPOINT  CALLS, ITER  ITERATIONS                *
*     2ND RUN (VEGAS1): IPOINT1 CALLS, ITER1 ITERATIONS                *
*     RESULT "RES" IS RETURNED WITH DESIRED ACCURACY "ACC"             *
*                                                                      *
************************************************************************



C----------------------------- MAIN PROGRAM ----------------------------
C
C     IMPLICIT REAL*8 (A-H,K-Z)
C
C     EXTERNAL FUNCT
C
C     COMMON/IOUT/IPRINT
C
C     ...
C
C     PARAMETERS FOR VEGAS
C     IDIM    =  ...
C     IPOINT  =  ...
C     ITER    =   10
C     IPOINT1 =  ...
C     ITER1   =    3
C     ACC     = 1.D-5
C
C     NO / LONG / SHORT PRINTOUT OF INTEGRALS: IPRINT = 0 / 1 / 10
C     IPRINT = 0
C
C     ...
C
C     CALL INTEG(FUNCT,IDIM,IPOINT,ITER,IPOINT1,ITER1,ACC,RES)
C
C     ...
C
C     STOP
C     END
C
C-----------------------------------------------------------------------




      SUBROUTINE INTEG(FXN,IDIM,IPOINT,ITER,IPOINT1,ITER1,ACC,RES)

      IMPLICIT REAL*8 (A-H,K-Z)

      COMMON/BVEG2/INDO,IT,SI,SI2,SWGT,SCHI,XI(50,10),SCALLS
     1,D(50,10),DI(50,10),INXI(50,10)
      COMMON/RESULT/S1,S2,S3,S4
      COMMON/IOUT/IPRINT

      EXTERNAL FXN
      IF (IPRINT.NE.0)
     .PRINT *,'------------------- CALL TO VEGAS -------------------'
      CALL VEGAS(FXN,ACC,IDIM,IPOINT,ITER,IPRINT,0)
      IF (IPRINT.NE.0)
     .PRINT *,'------------------- CALL TO VEGAS1 ------------------'
      CALL VEGAS1(FXN,ACC,IDIM,IPOINT1,ITER1,IPRINT,0)
      IF (IPRINT.NE.0) PRINT *,'        '

      RES=S1

      RETURN
      END




      SUBROUTINE VEGAS(FXN,BCC,NDIM,NCALL,ITMX,NPRN,IGRAPH)
      IMPLICIT REAL*8 (A-H,O-Z)
      REAL*4 UNIV
      COMMON/BVEG2/INDO,IT,SI,SI2,SWGT,SCHI,XI(50,10),SCALLS
     1,D(50,10),DI(50,10),INXI(50,10)
C***  ADDED BY RH (FOR SOME FORTRAN COMPILERS): 
      COMMON/BVEG3/ACC
C***  END OF ADDITION.
      DIMENSION XIN(50),R(50),DX(10),IA(10),KG(10),DT(10)
      DIMENSION XL(10),XU(10),QRAN(10),X(10)
      COMMON/RESULT/S1,S2,S3,S4
       EXTERNAL FXN
      DATA XL,XU/10*0.D0,10*1.D0/
      DATA NDMX/50/,ALPH/1.5D0/,ONE/1.D0/,MDS/1/
      IPR=1
      IF(NPRN.GT.0)IPR=0
      INDO=1
      DO 1 J=1,NDIM
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C1     XI(1,J)=ONE
         XI(1,J)=ONE
 1    CONTINUE
C===  END OF MODIFICATION
      ENTRY VEGAS1(FXN,BCC,NDIM,NCALL,ITMX,NPRN,IGRAPH)
      NOW=IGRAPH
C---  NEXT LINE TAKEN OUT ON 04/06/92 BY J. ZUNFT
C---  IF(IGRAPH.GT.0)CALL INPLOT(NOW,F1,W)
      IT=0
      SI=0.D0
      SI2=SI
      SWGT=SI
      SCHI=SI
      SCALLS=SI
      ENTRY VEGAS2(FXN,BCC,NDIM,NCALL,ITMX,NPRN,IGRAPH)
      ND=NDMX
      NG=1
      IF(MDS.EQ.0) GO TO 2
      NG=(NCALL*0.5D0)**(1.D0/NDIM)
      MDS=1
      IF((2*NG-NDMX).LT.0) GO TO 2
      MDS=-1
      NPG=NG/NDMX+1
      ND=NG/NPG
      NG=NPG*ND
2     K=NG**NDIM
      NPG=NCALL/K
      IF(NPG.LT.2)NPG=2
      CALLS=NPG*K
      DXG=ONE/NG
      DV2G=DXG**(2*NDIM)/NPG/NPG/(NPG-ONE)
      XND=ND
      NDM=ND-1
      DXG=DXG*XND
      XJAC=ONE
      DO 3 J=1,NDIM
      DX(J)=XU(J)-XL(J)
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C3     XJAC=XJAC*DX(J)
      XJAC=XJAC*DX(J)
 3    CONTINUE
C===  END OF MODIFICATION
      IF(ND.EQ.INDO) GO TO 8
      RC=INDO/XND
      DO 7 J=1,NDIM
      K=0
      XN=0.D0
      DR=XN
      I=K
4     K=K+1
      DR=DR+ONE
      XO=XN
      XN=XI(K,J)
5     IF(RC.GT.DR) GO TO 4
      I=I+1
      DR=DR-RC
      XIN(I)=XN-(XN-XO)*DR
      IF(I.LT.NDM) GO TO 5
      DO 6  I=1,NDM
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C6     XI(I,J)=XIN(I)
         XI(I,J)=XIN(I)
 6    CONTINUE
C===  END OF MODIFICATION
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C7     XI(ND,J)=ONE
      XI(ND,J)=ONE
 7    CONTINUE
C===  END OF MODIFICATION 
      INDO=ND
      ACC=BCC
8     IF(NPRN.NE.0.AND.NPRN.NE.10)PRINT 200,NDIM,CALLS,IT,ITMX
     1,ACC,MDS,ND
      IF(NPRN.EQ.10)PRINT 290,NDIM,CALLS,ITMX,ACC,MDS,ND
      ENTRY VEGAS3(FXN,BCC,NDIM,NCALL,ITMX,NPRN,IGRAPH)
9     IT=IT+1
      TI=0.D0
      TSI=TI
C---  NEXT LINE TAKEN OUT ON 04/06/92 BY J. ZUNFT
C---  IF(IGRAPH.GT.0)CALL REPLOT(NOW,F1,W)
      DO 10 J=1,NDIM
      KG(J)=1
      DO 101 I=1,ND
      INXI(I,J)=0
      D(I,J)=TI
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C10    DI(I,J)=TI
      DI(I,J)=TI
 101  CONTINUE
 10   CONTINUE
C===  END OF MODIFICATION
11    FB=0.D0
      F2B=FB
      K=0
12    K=K+1
      DO 121 J=1,NDIM
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C121   QRAN(J)=DBLE(UNIV())
      QRAN(J)=DBLE(UNIV())
 121  CONTINUE
C===  END OF MODIFICATION
      WGT=XJAC
      DO 15 J=1,NDIM
      XN=(KG(J)-QRAN(J))*DXG+ONE
      IA(J)=XN
      IAJ=IA(J)
      IAJ1=IAJ-1
      IF(IAJ.GT.1) GO TO 13
      XO=XI(IAJ,J)
      RC=(XN-IAJ)*XO
      GO TO 14
13    XO=XI(IAJ,J)-XI(IAJ1,J)
      RC=XI(IAJ1,J)+(XN-IAJ)*XO
14    X(J)=XL(J)+RC*DX(J)
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C15    WGT=WGT*XO*XND
      WGT=WGT*XO*XND
 15   CONTINUE
C===  END OF MODIFICATION
      F=FXN(X)*WGT
      F1=F/CALLS
      W=WGT/CALLS
C---  NEXT LINE TAKEN OUT ON 04/06/92 BY J. ZUNFT
C---  IF(IGRAPH.GT.0)CALL XPLOT(NOW,F1,W)
      F2=F*F
      FB=FB+F
      F2B=F2B+F2
      DO 16 J=1,NDIM
      IAJ=IA(J)
      INXI(IAJ,J)=INXI(IAJ,J)+1
      DI(IAJ,J)=DI(IAJ,J)+F/CALLS
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C16    IF(MDS.GE.0)  D(IAJ,J)=D(IAJ,J)+F2
      IF(MDS.GE.0)  D(IAJ,J)=D(IAJ,J)+F2
 16   CONTINUE
C===  END OF MODIFICATION
      IF(K.LT.NPG) GO TO 12
      F2B=F2B*NPG
      F2B=SQRT(F2B)
      F2B=(F2B-FB)*(F2B+FB)
      TI=TI+FB
      TSI=TSI+F2B
      IF(MDS.GE.0) GO TO 18
      DO 17 J=1,NDIM
      IAJ=IA(J)
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C17    D(IAJ,J)=D(IAJ,J)+F2B
      D(IAJ,J)=D(IAJ,J)+F2B
 17   CONTINUE
C===  END OF MODIFICATION
18    K=NDIM
19    KG(K)=MOD(KG(K),NG)+1
      IF(KG(K).NE.1) GO TO 11
      K=K-1
      IF(K.GT.0) GO TO 19
      TI=TI/CALLS
      TSI=TSI*DV2G
      TI2=TI*TI
C--------- CHANGE BY J. ZUNFT, 05/06/92------------------
      IF (DABS(TSI).LT.1.D-35) TSI = DSIGN(1.D-35,TSI)
C--------------------------------------------------------
      WGT=TI2/TSI
      SI=SI+TI*WGT
      SI2=SI2+TI2
      SWGT=SWGT+WGT
      SCHI=SCHI+TI2*WGT
      SCALLS=SCALLS+CALLS
C--------- CHANGE BY J. ZUNFT, 05/06/92------------------
      IF (DABS(SWGT).LT.1.D-35) SWGT = DSIGN(1.D-35,SWGT)
      IF (DABS(SI2) .LT.1.D-35) SI2  = DSIGN(1.D-35,SI2)
C--------------------------------------------------------
      AVGI=SI/SWGT
      SD=SWGT*IT/SI2
      CHI2A=0.
      IF(IT.GT.1)CHI2A=SD*(SCHI/SWGT-AVGI*AVGI)/(IT-1)
      SD=ONE/SD
C--------- CHANGE BY J. ZUNFT, 04/14/92-----------------
      IF (SD.LT.0.D0) THEN
      PRINT *,' SD  = ',SD,' < 0: SIGN CHANGED'
      SD = DABS(SD)
      END IF
      IF (TSI.LT.0.D0) THEN
      PRINT *,' TSI = ',TSI,' < 0: SIGN CHANGED'
      TSI = DABS(TSI)
      END IF
C-------------------------------------------------------
      SD=DSQRT(SD)
      IF(NPRN.EQ.0) GO TO 21
      TSI=DSQRT(TSI)
      IF(NPRN.NE.10)PRINT 201,IPR,IT,TI,TSI,AVGI,SD,CHI2A
      IF(NPRN.EQ.10)PRINT 203,IT,TI,TSI,AVGI,SD,CHI2A
      IF(NPRN.GE.0) GO TO 21
      DO 20 J=1,NDIM
      PRINT 202,J
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C20    PRINT 204,(XI(I,J),DI(I,J),D(I,J),I=1,ND)
      PRINT 204,(XI(I,J),DI(I,J),D(I,J),I=1,ND)
 20   CONTINUE
C===  END OF MODIFICATION
C--------- CHANGE BY J. ZUNFT, 05/06/92------------------
21    IF (DABS(AVGI).LT.1.D-35) AVGI = DSIGN(1.D-35,AVGI)
C--------------------------------------------------------
      IF(ABS(SD/AVGI).LE.ABS(ACC).OR.IT.GE.ITMX)NOW=2
      S1=AVGI
      S2=SD
      S3=TI
      S4=TSI
C---  NEXT LINE TAKEN OUT ON 04/06/92 BY J. ZUNFT
C---  IF(IGRAPH.GT.0)CALL PLOTIT(NOW,F1,W)
C      DO 23 J=1,NDIM
C      XO=D(1,J)
C      XN=D(2,J)
C      D(1,J)=(XO+XN)*0.5D0
C      DT(J)=D(1,J)
C      DO 22 I=2,NDM
C      D(I,J)=XO+XN
C      XO=XN
C      XN=D(I+1,J)
C      D(I,J)=(D(I,J)+XN)/3.D0
C22    DT(J)=DT(J)+D(I,J)
C      D(ND,J)=(XN+XO)*0.5D0
C23    DT(J)=DT(J)+D(ND,J)
C-----THIS PART OF THE VEGAS-ALGORITHM IS UNSTABLE
C-----IT SHOULD BE REPLACED BY
      DO 23 J=1,NDIM
      DT(J)=0.D0
      DO 231 I=1,ND
      IF(INXI(I,J).GT.0)D(I,J)=D(I,J)/INXI(I,J)
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C23    DT(J)=DT(J)+D(I,J)
      DT(J)=DT(J)+D(I,J)
 231  CONTINUE
 23   CONTINUE
C===  END OF MODIFICATION
      DO 28 J=1,NDIM
      RC=0.D0
      DO 241 I=1,ND
      R(I)=0.D0
C--------- CHANGE BY J. ZUNFT, 04/15/92 ---------
C---  IF(D(I,J).LE.0.D0)GO TO 24
      IF(D(I,J).EQ.0.D0.OR.DT(J)/D(I,J).LE.0.D0) GO TO 24
C------------------------------------------------
      XO=DT(J)/D(I,J)
C--------- CHANGE BY J. ZUNFT, 05/06/92 ---------
      IF (XO.EQ.1.D0) XO = 0.99999999999999999D0
C------------------------------------------------
      R(I)=((XO-ONE)/XO/DLOG(XO))**ALPH
24    RC=RC+R(I)
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
 241  CONTINUE
C===  END OF MODIFICATION
      RC=RC/XND
      K=0
      XN=0.D0
      DR=XN
      I=K
25    K=K+1
      DR=DR+R(K)
      XO=XN
      XN=XI(K,J)
26    IF(RC.GT.DR) GO TO 25
      I=I+1
      DR=DR-RC
C----------- CHANGE BY J. ZUNFT, 04/14/92 --------------
C---  XIN(I)=XN-(XN-XO)*DR/R(K)
      IF (DR.EQ.0.D0) THEN
      XIN(I)=XN
      ELSE
      XIN(I)=XN-(XN-XO)*DR/R(K)
      END IF
C-------------------------------------------------------
      IF(I.LT.NDM) GO TO 26
      DO 27 I=1,NDM
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C27    XI(I,J)=XIN(I)
      XI(I,J)=XIN(I)
 27   CONTINUE
C===  END OF MODIFICATION
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C28    XI(ND,J)=ONE
      XI(ND,J)=ONE
 28   CONTINUE
C===  END OF MODIFICATION
      IF(IT.LT.ITMX.AND.ABS(ACC).LT.ABS(SD/AVGI))GO TO 9
200   FORMAT(35H0INPUT PARAMETERS FOR VEGAS   NDIM=,I3
     1,8H  NCALL=,F8.0/28X,5H  IT=,I5,8H  ITMX =,I5/28X
     2,6H  ACC=,G9.3/28X,6H  MDS=,I3,6H   ND=,I4//)
290   FORMAT(13H0VEGAS  NDIM=,I3,8H  NCALL=,F8.0,8H  ITMX =,I5
     1,6H  ACC=,G9.3,6H  MDS=,I3,6H   ND=,I4)
201   FORMAT(/I1,20HINTEGRATION BY VEGAS/13H0ITERATION NO,I3,
     114H.   INTEGRAL =,G14.8/20X,10HSTD DEV  =,G10.4/
     234H ACCUMULATED RESULTS.   INTEGRAL =,G14.8/
     324X,10HSTD DEV  =,G10.4 / 24X,18HCHI**2 PER ITN   =,G10.4)
202   FORMAT(14H0DATA FOR AXIS,I2 / 7X,1HX,7X,10H  DELT I  ,
     12X,11H CONVCE    ,11X,1HX,7X,10H  DELT I  ,2X,11H CONVCE
     2,11X,1HX,7X,10H  DELT I  ,2X,11H CONVCE     /)
204   FORMAT(1X,3G12.4,5X,3G12.4,5X,3G12.4)
203   FORMAT(1H ,I3,G20.8,G12.4,G20.8,G12.4,G12.4)
C---  NEXT 3 LINES TAKEN OUT ON 04/06/92 BY J. ZUNFT
C---  S1=AVGI
C---  S2=SD
C---  S3=CHI2A
      RETURN
      END





C-----------------------------------------------------------------------
C   INITIALIZING THE RANDOM NUMBER GENERATOR
C     IF OLD CONFIGURATION EXISTS THEN
C     CALL UREAD(11)
C     ELSE
C     CALL RSTART(12,34,56,78)
C     END IF
C-----------------------------------------------------------------------


C----------------------------------------------------------------------
C  A UNIVERSAL RANDOM NUMBER GENERATOR

        FUNCTION UNIV()
        REAL U(97)
        COMMON /SET1/ U,C,CD,CM,I,J
        UNIV=U(I)-U(J)
        IF(UNIV.LT.0.) UNIV=UNIV+1.
        U(I)=UNIV
        I=I-1
        IF(I.EQ.0) I=97
        J=J-1
        IF(J.EQ.0) J=97
        C=C-CD
        IF(C.LT.0.) C=C+CM
        UNIV=UNIV-C
        IF(UNIV.LT.0.) UNIV=UNIV+1
        RETURN
        END

C----------------------------------------------------------------------
C SAVING THE RANDOM NUMBER GENERATOR CONFIGURATION

        SUBROUTINE UWRITE(IUNIT)
        INTEGER IUNIT
        REAL U(97)
        COMMON /SET1/ U,C,CD,CM,I,J
        OPEN(IUNIT,FILE='TESTRN',FORM='UNFORMATTED')
        WRITE(IUNIT) U,C,CD,CM,I,J
        CLOSE(IUNIT)
        RETURN
        END


C----------------------------------------------------------------------
C READING IN THE RANDOM NUMBER GENERATOR CONFIGURATION

        SUBROUTINE UREAD(IUNIT)
        INTEGER IUNIT
        REAL U(97)
        COMMON /SET1/ U,C,CD,CM,I,J
        OPEN(IUNIT,FILE='TESTRN',FORM='UNFORMATTED')
        READ(IUNIT) U,C,CD,CM,I,J
        CLOSE(IUNIT)
        RETURN
        END

C----------------------------------------------------------------------
C INITIALIZING THE RANDOM NUMBER GENERATOR
C TO INITIALIZE CALL RSTART(12,34,56,78)

        SUBROUTINE RSTART(I,J,K,L)
        REAL U(97)
        COMMON /SET1/ U,C,CD,CM,ISTART,JSTART
        IF ((I.LT.0).OR.(I.GT.178 )) STOP 'FIRST SEED .LT.0 OR .GT.178'
        IF ((J.LT.0).OR.(J.GT.178 )) STOP 'SECOND SEED .LT.0 OR .GT.178'
        IF ((K.LT.0).OR.(K.GT.178 )) STOP 'THIRD SEED .LT.0 OR .GT.178'
        IF ((L.LT.0).OR.(L.GT.168 )) STOP 'FOURTH SEED .LT.0 OR .GT.168'
        IF ( (I.EQ.1).AND.(J.EQ.1).AND.(K.EQ.1) ) STOP
     &     'FIRST, SECOND AND THIRD SEEDS ARE ALL EQUAL TO 1'
        ISTART=97
        JSTART=33
        IDUM = I
        JDUM = J
        KDUM = K
        LDUM = L
        DO 2 II=1,97
        S=0.
        T=.5
        DO 3 JJ=1,24
          M=MOD(MOD(IDUM*JDUM,179)*KDUM,179)
          IDUM=JDUM
          JDUM=KDUM
          KDUM=M
          LDUM=MOD(53*LDUM+1,169)
          IF(MOD(LDUM*M,64).GE.32) S=S+T
C===  NEXT LINES MODIFIED ON 2022/07/04 BY HEL TO MODERNIZE DO LOOP TERMINATION
C3         T=.5*T
          T=.5*T
 3        CONTINUE
C2         U(II)=S
          U(II)=S
 2        CONTINUE
C===  END OF MODIFICATION
        C=362436./16777216.
        CD=7654321./16777216.
        CM=16777213./16777216.
        RETURN
        END



C ======================================================================












